File: groupon-copycat-frontend/src/app/models/coupon.ts
/* tslint:disable:no-redundant-jsdoc */
import {CouponCategory} from '../enums/coupon-category.enum';
import {Company} from './company';

export class Coupon {
  private couponId: number;
  private amount: number;
  private price: number;
  private title: string;
  private description: string;
  private image: string;
  private startDate: Date;
  private endDate: Date;
  private category: CouponCategory;
  private company: Company;

  constructor($couponId: number, $amount: number, $price: number, $title: string, $description: string, $image: string
    , $startDate: Date, $endDate: Date, category: CouponCategory, company: Company) {
    this.couponId = $couponId;
    this.amount = $amount;
    this.price = $price;
    this.title = $title;
    this.description = $description;
    this.image = $image;
    this.startDate = $startDate;
    this.endDate = $endDate;
    this.category = category;
    this.company = company;
  }


  /**
   * Getter $couponId
   * @return {number}
   */
  public get $couponId(): number {
    return this.couponId;
  }

  /**
   * Setter $couponId
   * @param {number} value
   */
  public set $couponId(value: number) {
    this.couponId = value;
  }

  /**
   * Getter $amount
   * @return {number}
   */
  public get $amount(): number {
    return this.amount;
  }

  /**
   * Setter $amount
   * @param {number} value
   */
  public set $amount(value: number) {
    this.amount = value;
  }

  /**
   * Getter $price
   * @return {number}
   */
  public get $price(): number {
    return this.price;
  }

  /**
   * Setter $price
   * @param {number} value
   */
  public set $price(value: number) {
    this.price = value;
  }

  /**
   * Getter $title
   * @return {string}
   */
  public get $title(): string {
    return this.title;
  }

  /**
   * Setter $title
   * @param {string} value
   */
  public set $title(value: string) {
    this.title = value;
  }

  /**
   * Getter $description
   * @return {string}
   */
  public get $description(): string {
    return this.description;
  }

  /**
   * Setter $description
   * @param {string} value
   */
  public set $description(value: string) {
    this.description = value;
  }

  /**
   * Getter $image
   * @return {string}
   */
  public get $image(): string {
    return this.image;
  }

  /**
   * Setter $image
   * @param {string} value
   */
  public set $image(value: string) {
    this.image = value;
  }

  /**
   * Getter $startDate
   * @return {Date}
   */
  public get $startDate(): Date {
    return this.startDate;
  }

  /**
   * Setter $startDate
   * @param {Date} value
   */
  public set $startDate(value: Date) {
    this.startDate = value;
  }

  /**
   * Getter $endDate
   * @return {Date}
   */
  public get $endDate(): Date {
    return this.endDate;
  }

  /**
   * Setter $endDate
   * @param {Date} value
   */
  public set $endDate(value: Date) {
    this.endDate = value;
  }

  /**
   * Getter $couponCategory
   * @return {CouponCategory}
   */
  public get $category(): CouponCategory {
    return this.category;
  }

  /**
   * Setter $couponCategory
   * @param {CouponCategory} value
   */
  public set $category(value: CouponCategory) {
    this.category = value;
  }  /**
   * Getter $couponCategory
   * @return {CouponCategory}
   */
  public get $company(): Company {
    return this.company;
  }

  /**
   * Setter $couponCategory
   * @param {CouponCategory} value
   */
  public set $company(value: Company) {
    this.company = value;
  }

}

File: groupon-copycat/src/main/java/groupon_copycat/Spring_Boot/beans/Coupon.java
package groupon_copycat.Spring_Boot.beans;

import groupon_copycat.Spring_Boot.enums.CategoryType;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.*;
import java.sql.Date;

@Entity
@Table(name = "coupons")
@JsonIgnoreProperties(ignoreUnknown = true)
public class Coupon {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int couponId;
    @Column
    private int amount;
    @Column
    private double price;
    @Column
    private String title, description, image;
    @Column
    private Date startDate, endDate;
    @Column
    private CategoryType category;
    @ManyToOne
    @JoinColumn(name = "company_id", nullable = false)
//    @JsonIgnore
    private Company company;

    public Coupon(int amount, double price, String title, String description, String image, Date startDate,
                  Date endDate, CategoryType category /*, Company company*/) {
        this.amount = amount;
        this.price = price;
        this.title = title;
        this.description = description;
        this.image = image;
        this.startDate = startDate;
        this.endDate = endDate;
        this.category = category;
        //this.company = company;
    }

    public Coupon() { // ctor for Hibernate
    }

    @Override
    public String toString() {
        return "Coupon [couponId=" + couponId + ", amount=" + amount + ", price=" + price + ", title=" + title
                + ", description=" + description + ", image=" + image + ", startDate=" + startDate + ", endDate="
                + endDate + ", category=" + category + ", company=" + company + "]";
    }

    // for .contains()
    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Coupon) {
            return ((Coupon) obj).getCouponId() == this.getCouponId();
        }
        return false;
    }

    public int getAmount() {
        return amount;
    }

    public void setAmount(int amount) {
        this.amount = amount;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public Date getStartDate() {
        return startDate;
    }

    public void setStartDate(Date startDate) {
        this.startDate = startDate;
    }

    public Date getEndDate() {
        return endDate;
    }

    public void setEndDate(Date endDate) {
        this.endDate = endDate;
    }

    public int getCouponId() {
        return couponId;
    }

    public CategoryType getCategory() {
        return category;
    }

    public void setCategory(CategoryType category) {
        this.category = category;
    }

    public Company getCompany() {
        return company;
    }

    /**
     * <b>
     * <h1>DO NOT USE ANYWHERE OUTSIDE COMPANY FACADE</h1></b>
     * otherwise this can (if not used carefully) bypass some checks
     */
    public void setCompany(Company company) {
        this.company = company;
    }

}
